Calculate conflict markers instead of hardcoding

afresh1 opened the pull request:
It turns out that some tooling (Game of Trees) has special handling for conflict markers and will continually mark files that have them.

Since the test doesn't actually need them to be expanded, calculate them instead to not trigger the detection.

https://marc.gameoftrees.org/mail/1676484047.9764_0.html

afresh1 committed:
Calculate conflict markers instead of hardcoding

It turns out that some tooling (Game of Trees) has special handling for
conflict markers and will continually mark files that have them.

Since the test doesn't actually need them to be expanded,
calculate them instead to not trigger the detection.

afresh1 commented:
Apparently there is another instance that has less obvious workarounds in t/lib/croak/toke.  If I find a fix for that before this merges I'll push another commit.

afresh1 commented:
Both because I don't see an obvious "fix" for the the croak/toke test and because [the folks over at GoT have decided that having conflict markers in files might be necessary](https://marc.gameoftrees.org/mail/1676548247.31257_0.html), I am going to close this and leave it be.  I don't think the code is significantly improved by obfuscating those markers.

demerphq commented:
On Fri, 17 Feb 2023, 01:29 Andrew Fresh, ***@***.***> wrote:

> Closed #20813 <https://github.com/Perl/perl5/pull/20813>.
>

I'm inclined to say we should keep this open regardless, I don't think we
should have conflict markers in our files, even for tests. I haven't looked
closely yet at the croak test andrew got stuck on but the general intent of
this PR seemed pretty reasonable to me.

Yves

afresh1 commented:
That seems fine by me.  I will reopen and look forward to seeing ideas on fixing the croak/toke test.

demerphq commented:
I think the toke test depends on t/test.pl run_multiple_progs() (or maybe setup_multiple_progs) which needs to be patched to convert whatever replacement we use for conflict markers to the correct marker. Eg it could convert '<conflict<', '=conflict=' and '>conflict>' to the correct strings (and/or back, if necessary) it should work. I'm stuck on a phone just now so I may have overlooked something, but this is the general approach I'd take.

demerphq committed:
t/test.pl - support NOTE comments in code for run_multiple_progs()

These comments will be autoremoved from the code before it is executed.
*Must* start the line, and have exactly one space between the comment
char and the NOTE string, which must be uppercased.

This will be used in the following patch.

demerphq committed:
t/test.pl - support obfu conflict markers for code via run_multiple_progs()

This was we dont have to have literal conflict markers in test code
that is executed via run_multiple_progs(). For instance, t/lib/croak/toke
patch for that will follow.

demerphq committed:
t/lib/croak/toke - hide conflict markers in test code

This uses the previous two patches to add a NOTE comment
to the test for conflict markers explaining that we are using
the obfu style markers that run_multiple_progs() supports (now),
so that picky tools don't think we have conflict markers in
our files.

demerphq commented:
I updated the branch with code to fix the croak/toke issue.
